### DISCUSSION TITLE
A way to delay a callback until after the current transaction has committed successfully

### DISCUSSION THREAD

**Discussion by diesieben07** [payloadcms/payload#4118] [Feature Requests & Ideas]
It would be useful to have a way to delay code until after the current transaction is committed. The use-case for this are external effects that happen as a result of an "after change" hook or similar. I am currently adding push notifications to a backend using Payload, which should be sent whenever there is a new document in a specific collection. If I send out the notification in an "after change" hook, but then transaction rolls back, I have just sent out a notification for something that doesn't actually exist.

## Prior Art
- Django has [`transaction.on_commit`](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#performing-actions-after-commit)
- Spring has [`TransactionalEventListener`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html)
- Laravel/Eloquent ORM has `DB::afterCommit` (I couldn't find the docs about this, but here is a blog post: https://divinglaravel.com/better-management-of-database-transactions-in-laravel-8)

The way this could be implemented is to store an array `transactionCallbacks` in `PayloadRequest` and have `commitTransaction` execute them after committing.

**Comment by ssyberg** [comment]
Great timing, was also looking for similar behavior today. Another use case is where a side effect requires the mongo data to be written and available. In my particular use case I have a side effect that needs to be run after `create` but keeps failing because the record doesn't exist yet (I've hacked this by explicitly committing the transaction early, but that is brittle and defeats the purpose of the transaction).
